File: Loenn/entities/chains/chained_falling_block.lua
local drawableRectangle = require("structs.drawable_rectangle")
local fakeTilesHelper = require("helpers.fake_tiles")

local chainedFallingBlock = {}

chainedFallingBlock.name = "CommunalHelper/ChainedFallingBlock"
chainedFallingBlock.fieldInformation = {
    tiletype = {
        options = fakeTilesHelper.getTilesOptions(),
        editable = false
    },
    fallDistance = {
        minimumValue = 0,
        fieldType = "integer"
    }
}

function chainedFallingBlock.depth(room, entity)
    return entity.behind and 5000 or 0
end

chainedFallingBlock.placements = {
    name = "chained_falling_block",
    data = {
        width = 8,
        height = 8,
        tiletype = "3",
        climbFall = true,
        behind = false,
        fallDistance = 64,
        centeredChain = false,
        chainOutline = true,
        indicator = false,
        indicatorAtStart = false,
        chainTexture = "objects/CommunalHelper/chains/chain"
    }
}

local fakeTilesSpriteFunction = fakeTilesHelper.getEntitySpriteFunction("tiletype", false)

function chainedFallingBlock.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 8, entity.height or 8

    local sprites = fakeTilesSpriteFunction(room, entity)

    local fallDistance = entity.fallDistance or 16
    local rect = drawableRectangle.fromRectangle("line", x, y, width, height + fallDistance, {1, 1, 1, 0.5})
    rect.depth = 0
    table.insert(sprites, rect)

    return sprites
end

return chainedFallingBlock

File: Ahorn/entities/chains/ChainedFallingBlock.jl
module CommunalHelperChainedFallingBlock

using ..Ahorn, Maple
using Cairo

@mapdef Entity "CommunalHelper/ChainedFallingBlock" ChainedFallingBlock(
    x::Integer,
    y::Integer,
    width::Integer=Maple.defaultBlockWidth,
    height::Integer=Maple.defaultBlockWidth,
    tiletype::String="3",
    climbFall::Bool=true,
    behind::Bool=false,
    fallDistance::Integer=64,
    centeredChain::Bool=false,
    chainOutline::Bool=true,
    indicator::Bool=false,
    indicatorAtStart::Bool=false,
    chainTexture::String="objects/CommunalHelper/chains/chain",
)

const placements = Ahorn.PlacementDict(
    "Falling Block (Chained) (Communal Helper)" => Ahorn.EntityPlacement(
        ChainedFallingBlock,
        "rectangle",
        Dict{String, Any}(),
        Ahorn.tileEntityFinalizer,
    ),
)

Ahorn.editingOptions(entity::ChainedFallingBlock) = Dict{String, Any}(
    "tiletype" => Ahorn.tiletypeEditingOptions()
)

Ahorn.minimumSize(entity::ChainedFallingBlock) = 8, 8
Ahorn.resizable(entity::ChainedFallingBlock) = true, true

Ahorn.selection(entity::ChainedFallingBlock) = Ahorn.getEntityRectangle(entity)

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::ChainedFallingBlock, room::Maple.Room) 
    x, y = Ahorn.position(entity)
    w = Int(get(entity.data, "width", 8))
    h = Int(get(entity.data, "height", 8)) + Int(get(entity.data, "fallDistance", 64))

    save(ctx)

    set_dash(ctx, [0.6, 0.2])
    set_antialias(ctx, 1)
    set_line_width(ctx, 1)

    # width & height extend up and left for some reason.
    Ahorn.drawRectangle(ctx, x + 1, y + 1, w - 1, h - 1, (0.0, 0.0, 0.0, 0.0), (1.0, 1.0, 1.0, 0.5))

    restore(ctx)

    Ahorn.drawTileEntity(ctx, room, entity)
end

end
